src/Illuminate/Database/Query/Builder.php

```diff
@@ -1811,6 +1811,72 @@ protected function addDynamic($segment, $connector, $parameters, $index)
         $this->where(Str::snake($segment), '=', $parameters[$index], $bool);
     }
 
+    /**
+     * Slice where conditions at the given offset and add them to the query as a nested condition.
+     *
+     * @param  array  $whereSlice
+     * @return void
+     */
+    protected function groupWhereSlice($whereSlice)
+    {
+        $whereBooleans = collect($whereSlice)->pluck('boolean');
+
+        // Here we'll check if the given subset of where clauses contains any "or"
+        // booleans and in this case create a nested where expression. That way
+        // we don't add any unnecessary nesting thus keeping the query clean.
+        if ($whereBooleans->contains('or')) {
+            $this->wheres[] = $this->createNestedWhere(
+                $whereSlice, $whereBooleans->first()
+            );
+        } else {
+            $this->wheres = array_merge($this->wheres, $whereSlice);
+        }
+    }
+
+    /**
+     * Nest where conditions by slicing them at the given where count.
+     *
+     * @param  int|null  $originalWhereCount
+     * @return $this
+     */
+    public function nestWheres($originalWhereCount = null)
```

Comment:
Copy from `Eloquent/Builder` with first parameter removed since we are now in the Query class itself.
+ made public
+ slight logic adjustment for the default case of `$originalWhereCount = null` (nest *all* wheres).

Reply:
This method doesn't exists on `Eloquent/Builder` atm? Can you point out where it is?

Reply:
oops correction: `groupWhereSliceForScope` renamed to `groupWhereSlice` since the original function was aimed at scopes specifically.

Reply:
`addNewWheresWithinGroup(...)` renamed to `nestWheres(...)` to be more explicit